File: lib/linter.py
"""Linter di marca pre-pubblicazione — l'ultimo cancello prima che un post esca.

Regole (fonte: BRAND-IDENTITY_sheis_2026-08-03.json + .claude/skills/sheis-brand-core/
guardrails.json — QUESTO file non inventa regole, le implementa meccanicamente):

  1. Prezzi e cifre commerciali — mai un numero che valga come prezzo, sconto, listino,
     margine. SHEis non vende online e non mostra mai un prezzo pubblico (brand-core §7).
  2. Lessico da negozio, in OGNI lingua — shop/carrello/acquista/cart/tienda/panier/
     Warenkorb/koszyk/loja/متجر e affini. Il posizionamento è "portale ordini"/
     "area riservata", mai un negozio.
  3. «Metodo 29» in ogni grafia e parafrasi — firewall non negoziabile, trasversale,
     vale su OGNI contenuto SHEis senza eccezioni (guardrails.json §1).
  4. Claim numerici fuori dall'elenco documentato — solo "15 minuti di posa",
     "99% di origine naturale", "tre fasi YOUNIC", "senza ammoniaca" sono numeri
     che SHEis può dimostrare. Ogni altra cifra-claim (percentuali, "X giorni",
     "X anni di garanzia"...) è indimostrabile finché non arriva da Mauro.
  5. Claim assoluti/clinici non documentati e "100% naturale" (contraddice il dato
     reale: 99%).

Ogni violazione BLOCK porta la REGOLA e la FRASE ESATTA che l'ha fatta scattare —
è il requisito esplicito: "blocca e dice quale regola ha fermato quale frase".
Nessuna eccezione silenziosa: se un testo passa, `LintResult.ok` è True e basta.
"""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field

from .vincoli_brand import (PATTERN_NEGOZIO, negozio_eccezione, numero_documentato,
                            nega_il_canale, viola_firewall, CLAIM_VIETATI,
                            FIREWALL_PATTERN, PREZZO)

from re import escape as _re_escape

BLOCK, WARN = "BLOCK", "WARN"

# ⚠️ REGRESSIONE ⑥b (revisione avversariale 2026-08-03): un carattere
# invisibile o non-\w inserito DENTRO una parola vietata ("sh​op",
# "car​rello", "acqui\xa0sta") spezza il confine \b che tutte le regole
# usano — "sh" e "op" diventano due token separati, nessuno dei due combacia
# con "shop". Sia lo zero-width space (U+200B) sia il NBSP (U+00A0) sono
# non-\w per il modulo `re`: entrambi rompono la tokenizzazione allo stesso
# modo, quindi entrambi vanno RIMOSSI (non sostituiti con uno spazio, che
# lascerebbe comunque due token separati) prima di qualunque controllo.
CARATTERI_INVISIBILI = re.compile(
    "[" + "".join([
        "​",  # zero width space
        "‌",  # zero width non-joiner
        "‍",  # zero width joiner
        "⁠",  # word joiner
        "﻿",  # zero width no-break space / BOM
        "­",  # soft hyphen
        " ",  # non-breaking space — nessun uso legittimo nel copy SHEis
        "‎", "‏",  # left/right-to-left mark
    ]) + "]"
)


def _normalizza_testo(testo: str) -> str:
    """Rimuove caratteri invisibili/non-\\w usati per spezzare una parola
    vietata a metà, e normalizza le forme di compatibilità Unicode (es.
    varianti a larghezza intera). Va chiamata PRIMA di qualunque pattern."""
    t = CARATTERI_INVISIBILI.sub("", testo)
    return unicodedata.normalize("NFKC", t)


@dataclass
class Violazione:
    livello: str
    regola: str
    dettaglio: str
    frase: str = ""

    def __str__(self) -> str:
        punta = f'  →  "{self.frase}"' if self.frase else ""
        return f"[{self.livello}] {self.regola}: {self.dettaglio}{punta}"


@dataclass
class LintResult:
    ok: bool
    violazioni: list[Violazione] = field(default_factory=list)

    @property
    def bloccanti(self) -> list[Violazione]:
        return [v for v in self.violazioni if v.livello == BLOCK]

    def motivo_blocco(self) -> str:
        """Una riga in italiano, per sheis_pubblicazioni.motivo_blocco. Vuota se ok."""
        if not self.bloccanti:
            return ""
        prime = self.bloccanti[:3]
        resto = len(self.bloccanti) - len(prime)
        righe = [str(v) for v in prime]
        if resto > 0:
            righe.append(f"(+{resto} altra/e violazione/i)")
        return " · ".join(righe)

    def render(self) -> str:
        if not self.violazioni:
            return "  linter: OK — nessuna violazione"
        return "\n".join(f"  {v}" for v in self.violazioni)


# ---------------------------------------------------------------- 1. prezzi
PREZZO_PATTERNS = [
    (r"[€$£]\s?\d", "simbolo di valuta seguito da una cifra"),
    (r"\b\d+[.,]?\d*\s?(euro|eur|dollari|usd|dollars)\b", "importo esplicito in cifre"),
    # ⚠️ REGRESSIONE ⑥a (revisione avversariale 2026-08-03): "Il trattamento
    # costa duecento euro" passava indisturbato — il pattern sopra pretende
    # SEMPRE una cifra accanto alla valuta, ed è esattamente così che un
    # testo generato scriverebbe un prezzo per eleganza (in lettere). Invece
    # di inseguire ogni numerale italiano scritto per esteso (cento, mille,
    # duecentocinquanta, ...), si blocca la parola-valuta NUDA: SHEis non ha
    # mai un motivo legittimo di nominare "euro"/"dollari" in un contenuto —
    # non vende online, non mostra prezzi in nessuna forma (brand-core §7).
    (r"\b(euro|euros|dollari|dollar|dollars|sterline|pound|pounds|centesimi|cents)\b",
     "valuta nominata (anche in lettere, senza cifra accanto)"),
    (
        r"\b(sconto|sconti|scontato|scontata|scontati|promo|promozione|saldo|saldi|"
        r"listino|listini|price\s?list|tarifa|pricing|"
        r"margine|margini|ricarico|markup|descuento|discount)\b",
        "lessico prezzo/sconto/listino/margine — SHEis non mostra mai un prezzo pubblico",
    ),
]

# ⚠️ La fonte elenca fra le cifre commerciali vietate anche «offerta»,
# «promo» e «saldo». Questa lista, scritta a mano, non le aveva: «Scrivici
# per ricevere un'offerta commerciale» passava di qui e veniva bloccata
# dagli altri due filtri. I termini della fonte si aggiungono, non si
# ricopiano — se domani ne arriva un altro, arriva da solo.
PREZZO_PATTERNS += [
    (r"\b" + _re_escape(t) + r"\w{0,3}\b", f"termine commerciale vietato dalla fonte: «{t}»")
    for t in PREZZO if t.isalpha() and len(t) >= 5
]
PREZZO_NUDO = [
    (r"\bprezzo|prezzi|precio|precios|price|prices\b", "parola 'prezzo' nominata"),
]

# ------------------------------------------------- 2. lessico da negozio (multilingua)
#
# Questa lista NON si scrive più qui. Era ricopiata a mano da BRAND-IDENTITY e
# aveva perso per strada spagnolo, portoghese e polacco declinato: «carrito»,
# «carrinho» e «koszyka» passavano da questo filtro mentre il gemello
# dell'outreach li bloccava — due verdetti opposti sullo stesso testo, e nessuno
# dei due che sembrasse rotto. Ora arriva generata dalla fonte.
NEGOZIO_PATTERNS = PATTERN_NEGOZIO
# Eccezione dichiarata: SHEis usa legittimamente "non vendiamo online" per NEGARE
# l'e-commerce — è la leva, non la violazione. Vale sullo stesso pattern dell'outreacher.
NEGOZIO_AMMESSO = [
    r"non (siamo|vendiamo) (in vendita )?online",
    r"nessun e-?commerce", r"niente e-?commerce",
    r"we don'?t sell online", r"no online sales",
    r"no vendemos online", r"no estamos en venta online",
]

# ---------------------------------------------------------- 3. firewall Metodo 29
# Trasversale, sempre BLOCK, nessuna eccezione — vedi guardrails.json forbidden_pairs.
# ⚠️ Il firewall NON si scrive più qui. Il collaudo del 3/8 ha misurato due
# vie d'uscita da questo elenco: «il metodo-29» col trattino e «approccio
# ventinove passi, applicato al metodo». Un elenco di stringhe non difende
# una regola che il cliente ha dichiarato non negoziabile. Le forme vengono
# dalla fonte e valgono per tutti e quattro i filtri insieme.
M29_PATTERNS = [(f['pattern'], f['cosa']) for f in FIREWALL_PATTERN]

# --------------------------------------------- 4/5. claim numerici e assoluti
# Whitelist ESATTA (brand-identity.regole_di_generazione.numeri_ammessi):
#   "15 minuti di posa" · "99% di origine naturale" · "tre fasi YOUNIC" · "senza ammoniaca"
CLAIM_QUANTIFICATO = [
    (r"\b\d{1,3}\s?%", "percentuale non nell'elenco documentato"),
    (r"\b\d+\s*(minuti|minuto|ore|ora)\b", "durata quantificata non documentata"),
    (r"\b\d+\s*(giorni|giorno|mesi|mese|anni|anno)\b", "quantità temporale non documentata (garanzia/esperienza)"),
    # ⚠️ Buco misurato il 3/8 sulla batteria d'insieme: «Con le nostre 120 nuance
    # disponibili» passava da TUTTI i filtri. La cartella SHEis Color ne ha 83 —
    # un numero sbagliato sull'unico dato di prodotto che il mercato verifica in
    # due secondi. Non era una divergenza fra filtri: era un buco condiviso, che
    # solo una prova d'insieme poteva far vedere.
    (r"\b\d+\s*(nuance|nuances|tonalit\w*|tonos|matices|shades)\b",
     "ampiezza di gamma non documentata (la cartella SHEis Color ha 83 nuance)"),
    (r"\b\d+\s*(client[ie]|salon[ie]|distributor[ie]|paesi|mercati)\b",
     "quantità commerciale non documentata"),
]
# La whitelist era ricopiata a mano anche qui, in una terza forma ancora
# diversa. Ora viene dalla fonte: `numero_documentato` legge NUMERI_DOCUMENTATI
# generato da BRAND-IDENTITY. Aggiungere un numero documentato si fa lì, una
# volta sola, e vale per tutti e quattro i filtri insieme.

# ⚠️ Anche i claim vengono dalla fonte. Scritti a mano lasciavano
# passare i numeri in LETTERE: «YOUNIC lavora in cinque fasi» passava
# da tutti e quattro i filtri, e le fasi documentate sono TRE.
CLAIM_ASSOLUTO_PATTERNS = [(c['pattern'], c['cosa']) for c in CLAIM_VIETATI]


def _find(testo: str, patterns, livello: str, regola: str) -> list[Violazione]:
    out = []
    for pat, dettaglio in patterns:
        m = re.search(pat, testo, re.IGNORECASE)
        if m:
            out.append(Violazione(livello, regola, dettaglio, m.group(0).strip()))
    return out


def _negozio_ammesso(testo: str, frase: str) -> bool:
    # Prima: la parola è innocente? Il confronto per radice prende «ordinario»
    # insieme a «ordina» — le eccezioni sono dichiarate nella fonte, non qui.
    if negozio_eccezione(frase):
        return True
    # Poi: la frase NEGA il canale? È testo approvato, non una violazione.
    if nega_il_canale(testo):
        return True
    basso = testo.lower()
    idx = basso.find(frase.lower())
    finestra = basso[max(0, idx - 60): idx + len(frase) + 25]
    return any(re.search(p, finestra) for p in NEGOZIO_AMMESSO)


def _claim_ammesso(testo: str, span: tuple[int, int]) -> bool:
    return numero_documentato(testo, span[0], span[1])


def lint_pubblicazione(testo: str, canale: str = "generico") -> LintResult:
    """Linter completo su un testo destinato alla pubblicazione (caption, copy
    secondario, alt-text). Va chiamato SEMPRE prima di mettere in coda un post,
    anche in DRY-RUN — il linter non è un effetto collaterale della pubblicazione
    vera, è un controllo indipendente da essa.
    """
    if not testo or not testo.strip():
        return LintResult(ok=False, violazioni=[Violazione(BLOCK, "vuoto", "testo vuoto o assente")])

    testo = _normalizza_testo(testo)
    v: list[Violazione] = []
    v += _find(testo, PREZZO_PATTERNS, BLOCK, "prezzi-e-cifre-commerciali")
    v += _find(testo, PREZZO_NUDO, WARN, "prezzo-nominato")
    v += _find(testo, M29_PATTERNS, BLOCK, "firewall-metodo-29")
    # ⚠️ I claim ora includono le quantità scritte in LETTERE («cinque fasi»),
    # e fra quelle c'è anche «tre fasi», che è documentata. Quindi anche questo
    # controllo deve consultare l'elenco dei numeri leciti, esattamente come fa
    # quello sulle cifre: senza, il filtro bloccherebbe il dato vero del
    # cliente — l'errore che questa mattina abbiamo già pagato sul «99%».
    for pat, dettaglio in CLAIM_ASSOLUTO_PATTERNS:
        m = re.search(pat, testo, re.IGNORECASE)
        if m and not _claim_ammesso(testo, m.span()):
            v.append(Violazione(BLOCK, "claim-non-documentato", dettaglio, m.group(0).strip()))

    for viol in _find(testo, NEGOZIO_PATTERNS, BLOCK, "lessico-da-negozio"):
        if not _negozio_ammesso(testo, viol.frase):
            v.append(viol)

    # claim numerici: BLOCK solo se la cifra NON è nel perimetro whitelisted
    for pat, dettaglio in CLAIM_QUANTIFICATO:
        for m in re.finditer(pat, testo, re.IGNORECASE):
            if not _claim_ammesso(testo, m.span()):
                v.append(Violazione(BLOCK, "claim-numerico-non-documentato", dettaglio, m.group(0).strip()))

    ok = not any(x.livello == BLOCK for x in v)
    return LintResult(ok=ok, violazioni=v)

File: src/lib/linter.ts
import { BRAND, GUARDRAILS } from "@/lib/brand";
import {
  FORME_FLESSE,
  NEGAZIONI_AMMESSE,
  CLAIM_VIETATI,
  ECCEZIONI_RADICE,
  NUMERI_DOCUMENTATI,
  negozioEccezione,
  numeroDocumentato,
} from "@/lib/vincoli-brand";

const ELENCO_DOCUMENTATI = NUMERI_DOCUMENTATI.map((n) => n.valore).join(", ");

/**
 * Il linter blocca, non avvisa (SPEC.md §"Vincoli non negoziabili" #2). Viene
 * chiamato in due soli punti: prima di salvare un contenuto come approvato, e
 * prima di metterlo in coda di pubblicazione. Ogni blocco dice QUALE regola
 * ha fermato QUALE frase, in italiano — mai un rifiuto generico.
 *
 * Regole 1-3 (prezzi, lessico da negozio, firewall Metodo 29) sono
 * deterministiche su liste misurate/validate dal cliente. Le regole 4-5
 * (claim numerici, nomi senza consenso) sono euristiche best-effort — lo
 * dichiarano esplicitamente i commenti sotto — perché non esiste ancora un
 * registro dei consensi in database: è un gap reale, non nascosto.
 */

export type ViolazioneLinter = {
  regola: string;
  descrizione: string;
  frase: string;
};

export type EsitoLinter = {
  bloccato: boolean;
  violazioni: ViolazioneLinter[];
};

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function estraiContesto(testo: string, indice: number, lunghezza: number): string {
  const inizio = Math.max(0, indice - 30);
  const fine = Math.min(testo.length, indice + lunghezza + 30);
  const prefisso = inizio > 0 ? "…" : "";
  const suffisso = fine < testo.length ? "…" : "";
  return `${prefisso}${testo.slice(inizio, fine).trim()}${suffisso}`;
}

function cercaTermini(testo: string, termini: readonly string[], regola: string, descrizione: string): ViolazioneLinter[] {
  const trovate: ViolazioneLinter[] = [];
  for (const termine of termini) {
    if (!termine) continue;
    // "€" e simili non hanno confini di parola: gestiti senza \b.
    // ⚠️ Il confine di parola `\b` è ASCII anche con la bandiera `u`: fra due
    // caratteri arabi non esiste, quindi «متجر» dentro «متجرنا» non veniva
    // mai trovato. Il confine ha senso solo per gli alfabeti latini.
    const haWordBoundary = /[A-Za-z0-9]/.test(termine);
    // ⚠️ Confronto per RADICE, non per parola esatta — misurato il 2026-08-03.
    // Con i soli confini di parola, «koszyka» (genitivo polacco di «koszyk»)
    // passava indisturbato: le lingue slave declinano e quelle romanze concordano,
    // quindi chi confronta parole intere blocca il caso da manuale e lascia passare
    // quello vero. Due linter dello stesso sistema davano verdetti diversi sullo
    // stesso testo, che è il modo peggiore di sbagliare.
    // Il suffisso libero è ammesso SOLO da 6 caratteri in su: sotto quella soglia
    // causerebbe falsi positivi grossolani ("cart" colpirebbe "cartella",
    // "compra" colpirebbe "comprare" — che però è già in elenco per esteso).
    const ammettiSuffisso = haWordBoundary && termine.length >= 6 && !termine.includes(" ");
    const pattern = !haWordBoundary
      ? new RegExp(escapeRegExp(termine), "gi")
      : ammettiSuffisso
        ? new RegExp(`\\b${escapeRegExp(termine)}\\p{L}{0,3}\\b`, "giu")
        : new RegExp(`\\b${escapeRegExp(termine)}\\b`, "giu");
    const m = pattern.exec(testo);
    if (m) {
      // Il confronto per radice ha un costo: «ordina» + 3 lettere prende anche
      // «ordinario». Le parole innocenti sono dichiarate nella fonte (§lessico.
      // eccezioni_radice) e valgono per tutti e quattro i filtri allo stesso modo.
      if (negozioEccezione(m[0])) continue;
      trovate.push({
        regola,
        descrizione: `${descrizione} — termine vietato: "${termine}"`,
        frase: estraiContesto(testo, m.index, m[0].length),
      });
    }
  }
  return trovate;
}

/** Regola 1 — prezzi e cifre commerciali. */
function checkPrezzi(testo: string): ViolazioneLinter[] {
  return cercaTermini(
    testo,
    BRAND.lessico.vietato_assoluto.prezzi_e_cifre_commerciali,
    "prezzi_cifre_commerciali",
    "Prezzo o cifra commerciale non ammessi in un contenuto SHEis",
  );
}

/** Regola 2 — lessico da negozio, in ogni lingua. */
const NEGAZIONI_RE = NEGAZIONI_AMMESSE.map((p) => new RegExp(p, "iu"));

/**
 * La frase NEGA il canale invece di proporlo?
 *
 * ⚠️ «Non siamo in vendita online, né Amazon né e-commerce nostro» è testo
 * APPROVATO e usato nel copione: nega il canale, è la leva di SHEis. Questo
 * filtro era l'UNICO dei quattro senza questa guardia, e lo bloccava perché
 * conteneva «e-commerce». Un filtro che rifiuta il testo approvato viene
 * disattivato da chi lo usa, e allora non ferma più niente.
 */
function negaIlCanale(testo: string): boolean {
  return NEGAZIONI_RE.some((r) => r.test(testo));
}

function checkLessicoNegozio(testo: string): ViolazioneLinter[] {
  if (negaIlCanale(testo)) return [];
  return cercaTermini(
    testo,
    [...BRAND.lessico.vietato_assoluto.lessico_da_negozio, ...FORME_FLESSE],
    "lessico_da_negozio",
    "Lessico da e-commerce/negozio non ammesso — il pubblico è sempre professionale",
  );
}

/** Regola 3 — firewall Metodo 29 (assoluto, in ogni grafia/lingua/parafrasi). */
function checkFirewallM29(testo: string): ViolazioneLinter[] {
  const alias = GUARDRAILS.worlds.world_m29.entities.flatMap((e) => [e.canonical, ...e.aliases]);
  const dirette = cercaTermini(
    testo,
    alias,
    "firewall_metodo_29",
    'Firewall Metodo 29: nessun collegamento pubblico ammesso con SHEis (regola non negoziabile)',
  );
  if (dirette.length > 0) return dirette;

  // Parafrasi/numero (test avversariale T13 in tests/firewall-m29.md): "ventinove"
  // o "29" in prossimità della parola "metodo" resta un indizio ricostruibile.
  const euristiche = [
    /\bventinove\w*\b[^.!?]{0,40}\bmetodo\b/giu,
    /\bmetodo\b[^.!?]{0,40}\bventinove\w*\b/giu,
    /\b29\b[^.!?]{0,30}\bmetodo\b/giu,
    /\bmetodo\b[^.!?]{0,30}\b29\b/giu,
  ];
  const trovate: ViolazioneLinter[] = [];
  for (const pattern of euristiche) {
    const m = pattern.exec(testo);
    if (m) {
      trovate.push({
        regola: "firewall_metodo_29_parafrasi",
        descrizione:
          "Firewall Metodo 29: formulazione che permette di ricostruire il collegamento (numero/parafrasi vicino a \"metodo\") — bloccata per prudenza, verifica a mano",
        frase: estraiContesto(testo, m.index, m[0].length),
      });
      break; // una sola segnalazione euristica basta a bloccare e a farsi rivedere
    }
  }
  return trovate;
}

/** Regola 4 — claim numerici non documentati (best-effort: vedi commento sopra). */
function checkNumeriNonDocumentati(testo: string): ViolazioneLinter[] {
  // ⚠️ Difetto misurato il 3/8: la whitelist era un insieme di CIFRE NUDE, senza
  // contesto — quindi «99% di sconto» e «15 anni di esperienza» passavano di qui
  // e venivano bloccati dagli altri tre filtri. A decidere è il contesto, non la
  // cifra. La regola vive ora in un solo posto: NUMERI_DOCUMENTATI, generato da
  // BRAND-IDENTITY, che tutti e quattro leggono.
  //
  // ⚠️ Il `\b` finale NON può applicarsi a "%": misurato il 2026-08-03.
  // "%" non è un carattere di parola, quindi `%\b` pretende che subito dopo ci sia
  // una lettera — cioè intercetta "100%naturale" e lascia passare "100% naturale",
  // che è il modo in cui chiunque lo scrive davvero. Il caso reale sfuggiva, quello
  // da manuale veniva bloccato. Ora "%" sta in un ramo senza `\b`, e le parole
  // conservano il loro confine.
  const pattern = /\b(\d+(?:[.,]\d+)?)\s*(?:(%|percento|per\s*cento)|(minuti|minuto|fasi|fase|clienti|cliente|follower|anni|anno|ore|ora|giorni|giorno|nuance|tonalità)\b)/giu;
  const trovate: ViolazioneLinter[] = [];
  let m: RegExpExecArray | null;
  while ((m = pattern.exec(testo)) !== null) {
    if (numeroDocumentato(testo, m.index, m.index + m[0].length)) continue;
    const unita = m[2] ?? m[3] ?? "";
    trovate.push({
      regola: "numero_non_documentato",
      descrizione: `Numero "${m[1]}" accostato a "${unita}" non è fra i dati documentati (${ELENCO_DOCUMENTATI}) — non basta la cifra, serve il contesto giusto: «99% di origine naturale» sì, «99% di sconto» no. Marca [DA CONFERMARE] o rimuovi`,
      frase: estraiContesto(testo, m.index, m[0].length),
    });
  }
  return trovate;
}

/** Regola 6 — claim assoluti, clinici e garanzie.
 *
 * ⚠️ Regola MANCANTE fino al 2026-08-03, trovata confrontando questo linter con
 * quello dei worker: «BABILON è una linea 100% naturale» passava qui e veniva
 * bloccato là. Due filtri dello stesso sistema con verdetti opposti sullo stesso
 * testo sono peggio di un filtro assente, perché nessuno dei due sembra rotto.
 *
 * Perché sono vietati, al di là della coerenza fra strumenti: sono affermazioni
 * che l'azienda non può dimostrare. «99% di origine naturale» è documentato e
 * passa; «100% naturale» non lo è. In cosmetica un claim di questo tipo non è
 * una sfumatura di marketing — è un'affermazione contestabile.
 */
const CLAIM_ASSOLUTI: readonly { re: RegExp; cosa: string }[] = [
  { re: /\b100\s*%/giu,                                                    cosa: "«100%» — nessun dato documentato lo sostiene (il valore verificato è 99% di origine naturale)" },
  { re: /\bclinicamente\s+(provat|testat|dimostrat)\w*/giu,                cosa: "claim clinico" },
  { re: /\bdermatologicamente\s+provat\w*/giu,                             cosa: "claim clinico (il documentato è «dermatologicamente testato», non «provato»)" },
  { re: /\brisultat\w*\s+garantit\w*/giu,                                  cosa: "garanzia di risultato" },
  { re: /\bgarantiam\w+\b/giu,                                             cosa: "garanzia di risultato" },
  // ⚠️ Servono le parole in mezzo: «il migliore PRODOTTO del mercato» è la forma
  // naturale, «il migliore del mercato» quella da manuale. Senza il tratto libero
  // si intercetta solo la seconda — lo stesso errore del "%" qui sopra, dove il
  // caso reale sfuggiva e quello di scuola veniva preso.
  { re: /\bmiglior\w*\b[^.!?]{0,30}\b(del|sul)\s+mercato/giu,               cosa: "superlativo assoluto non dimostrabile" },
  // «più» si scrive con l'accento (U+00F9), non con l'apostrofo: cercare `piu'`
  // avrebbe intercettato solo chi lo digita male. Si accettano entrambe le forme.
  { re: /\b(?:il|la)\s+pi(?:ù|u['’])\s+[^.!?]{0,30}\b(?:del|sul)\s+mercato/giu, cosa: "superlativo assoluto non dimostrabile" },
  { re: /\b(numero\s*uno|n\.?\s*1)\s+(in|del|sul)\b/giu,                   cosa: "primato non dimostrabile" },
  { re: /\bl['’]?unic\w+\s+(prodotto|linea|marchio|azienda)\b/giu,         cosa: "esclusività non dimostrabile" },
  { re: /\bcura\s+(la|il|i|le)\s+\w+/giu,                                  cosa: "possibile claim terapeutico — un cosmetico non cura" },
  { re: /\b(senza\s+alcun|zero)\s+(effett\w+\s+collateral\w+|rischi\w*)/giu, cosa: "assenza assoluta di rischi" },
];

function checkClaimAssoluti(testo: string): ViolazioneLinter[] {
  const trovate: ViolazioneLinter[] = [];
  // ⚠️ I claim della fonte includono le quantità scritte in LETTERE: «YOUNIC
  // lavora in cinque fasi» passava da questo filtro, e le fasi documentate sono
  // TRE. Si consulta comunque l'elenco dei numeri leciti, altrimenti si
  // bloccherebbe anche «tre fasi», che è il dato vero.
  for (const { pattern, cosa } of CLAIM_VIETATI) {
    const re = new RegExp(pattern, "iu");
    const m = re.exec(testo);
    if (m && !numeroDocumentato(testo, m.index, m.index + m[0].length)) {
      trovate.push({
        regola: "claim_non_documentato",
        descrizione: `Claim non dimostrabile: ${cosa} — riformula su un dato documentato o marca [DA CONFERMARE]`,
        frase: estraiContesto(testo, m.index, m[0].length),
      });
    }
  }
  for (const { re, cosa } of CLAIM_ASSOLUTI) {
    re.lastIndex = 0;
    const m = re.exec(testo);
    if (m) {
      trovate.push({
        regola: "claim_assoluto_o_clinico",
        descrizione: `Claim non dimostrabile: ${cosa} — riformula su un dato documentato o marca [DA CONFERMARE]`,
        frase: estraiContesto(testo, m.index, m[0].length),
      });
    }
  }
  return trovate;
}

/** Regola 5 — nomi di clienti/distributori senza consenso (denylist esplicita, oggi vuota: nessun nome è stato confermato pubblicabile). */
function checkNomiSenzaConsenso(testo: string, nomiSenzaConsenso: readonly string[]): ViolazioneLinter[] {
  if (nomiSenzaConsenso.length === 0) return [];
  return cercaTermini(
    testo,
    nomiSenzaConsenso,
    "nome_senza_consenso",
    "Nome di cliente/distributore citato senza consenso scritto registrato",
  );
}

export type OpzioniLint = {
  /** Nomi propri che NON hanno ancora consenso scritto a essere citati pubblicamente. */
  nomiSenzaConsenso?: readonly string[];
};

/** Punto unico di lint: concatena i campi pubblici e applica tutte le regole. */
export function lintContenuto(
  campi: { hook?: string; copy?: string; copySecondario?: string; cta?: string; hashtag?: readonly string[] },
  opzioni: OpzioniLint = {},
): EsitoLinter {
  const testo = [campi.hook, campi.copy, campi.copySecondario, campi.cta, (campi.hashtag ?? []).join(" ")]
    .filter(Boolean)
    .join("\n");

  const violazioni: ViolazioneLinter[] = [
    ...checkPrezzi(testo),
    ...checkLessicoNegozio(testo),
    ...checkFirewallM29(testo),
    ...checkNumeriNonDocumentati(testo),
    ...checkClaimAssoluti(testo),
    ...checkNomiSenzaConsenso(testo, opzioni.nomiSenzaConsenso ?? []),
  ];

  return { bloccato: violazioni.length > 0, violazioni };
}

/** Lint di un testo grezzo (usato anche fuori dal modello Contenuto, es. articoli sito). */
export function lintTesto(testo: string, opzioni: OpzioniLint = {}): EsitoLinter {
  const violazioni: ViolazioneLinter[] = [
    ...checkPrezzi(testo),
    ...checkLessicoNegozio(testo),
    ...checkFirewallM29(testo),
    ...checkNumeriNonDocumentati(testo),
    ...checkClaimAssoluti(testo),
    ...checkNomiSenzaConsenso(testo, opzioni.nomiSenzaConsenso ?? []),
  ];
  return { bloccato: violazioni.length > 0, violazioni };
}
